### PR TITLE
[SCEV] Retain SCEVSequentialMinMaxExpr if an operand may trigger UB.

### PR DESCRIPTION
**Pull request by fhahn**
Retain SCEVSequentialMinMaxExpr if an operand may trigger UB, e.g. if there is an UDiv operand that may divide by 0 or poison

### CODE REVIEW

**Comment by fhahn** [comment]
Ping :)

File: llvm/include/llvm/Analysis/ScalarEvolutionExpressions.h [hunk 1]
```diff
@@ -511,8 +511,9 @@ class SCEVUMinExpr : public SCEVMinMaxExpr {
 /// This node is the base class for sequential/in-order min/max selections.
 /// Note that their fundamental difference from SCEVMinMaxExpr's is that they
 /// are early-returning upon reaching saturation point.
-/// I.e. given `0 umin_seq poison`, the result will be `0`,
-/// while the result of `0 umin poison` is `poison`.
+/// I.e. given `0 umin_seq poison`, the result will be `0`, while the result of
+/// `0 umin poison` is `poison`. When returning early, later expressions are not
+/// executed.
```
Comment (nikic) [inline llvm/include/llvm/Analysis/ScalarEvolutionExpressions.h:516]:
```suggestion
/// executed, so `0 umin_seq (%x u/ 0)` does not result in undefined behavior.
```

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 2]
```diff
@@ -4304,6 +4304,16 @@ ScalarEvolution::getSequentialMinMaxExpr(SCEVTypes Kind,
   }
 
   for (unsigned i = 1, e = Ops.size(); i != e; ++i) {
+    bool MayBeUB = SCEVExprContains(Ops[i], [this](const SCEV *S) {
+      auto *UDiv = dyn_cast<SCEVUDivExpr>(S);
+      // The UDiv may be UB if the divisor is poison or zero. Unless the divisor
+      // is a non-zero constant, we have to assume the UDiv may be UB.
+      return UDiv && (!isa<SCEVConstant>(UDiv->getOperand(1)) ||
+                      !isKnownNonZero(UDiv->getOperand(1)));
```
Comment (nikic) [inline llvm/lib/Analysis/ScalarEvolution.cpp:4312]: As it looks like we're going to need it anyway, I think I'd add a isGuaranteedNotToBePoison() method (as a thin wrapper around SCEVPoisonCollector) in this PR and write this as `!isKnownNonZero || !isGuaranteedNotToBePoison`. I find the current way this is checked pretty confusing (at least it should mention that the reason for the SCEVConstant check is poison).

Comment (fhahn) [reply llvm/lib/Analysis/ScalarEvolution.cpp]: Updated thanks! Forgot about the poison visitor, used it as suggested.

File: llvm/include/llvm/Analysis/ScalarEvolutionExpressions.h (hunk 1 above)
Comment (fhahn) [reply llvm/include/llvm/Analysis/ScalarEvolutionExpressions.h]: Done, thanks!

**Comment by fhahn** [comment]
ping :)

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 3]
```diff
@@ -7299,6 +7309,11 @@ bool ScalarEvolution::isGuaranteedToTransferExecutionTo(const Instruction *A,
   return false;
 }
 
+bool ScalarEvolution::isGuaranteedNotToBePoison(const SCEV *Op) {
+  SCEVPoisonCollector PC(/* LookThroughMaybePoisonBlocking */ false);
```
Comment (nikic) [inline llvm/lib/Analysis/ScalarEvolution.cpp:7313]:
```suggestion
  SCEVPoisonCollector PC(/* LookThroughMaybePoisonBlocking */ true);
```
I think? If we have `%a umin_seq %b` where `%b` may be poison, the whole expression may be poison (if `%a != 0`).

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Analysis/ScalarEvolution.cpp (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Analysis/ScalarEvolution.cpp]: Yep, updated, thanks!

**Pull request merged**